File: CRM-Client/src/schemas/customerProfile.ts
import { z } from 'zod'

const JsonRecordSchema = z.record(z.unknown())

export const CustomerProfileFreshnessSchema = z.object({
  profile_as_of: z.string().nullable(),
  latest_business_event_at: z.string().nullable(),
  is_stale: z.boolean(),
  stale_reason: z.string().nullable()
})

export const CustomerProfileSectionsSchema = z.object({
  current_situation: JsonRecordSchema,
  current_journeys: z.array(JsonRecordSchema),
  important_changes: z.array(JsonRecordSchema),
  long_term_context: JsonRecordSchema,
  follow_up_process: z.array(JsonRecordSchema),
  recorded_follow_ups: z.array(JsonRecordSchema)
})

export const CustomerProfileEvidenceRefSchema = z.object({
  evidence_key: z.string().optional(),
  evidence_id: z.string().optional(),
  source_type: z.string().optional(),
  source_id: z.union([z.number().int(), z.string()]).nullable().optional(),
  source_version: z.union([z.number().int(), z.string()]).nullable().optional()
}).passthrough()

export const CustomerProfileEvidenceSchema = z.object({
  evidence_key: z.string(),
  source_type: z.string(),
  source_id: z.string().nullable(),
  source_version: z.union([z.number().int(), z.string()]).nullable(),
  occurred_at: z.string().nullable(),
  title: z.string().nullable(),
  snippet: z.string().nullable(),
  visibility: z.enum(['VISIBLE', 'UNAVAILABLE']),
  availability: z.enum(['AVAILABLE', 'UNAVAILABLE']),
  link: z.string().nullable(),
  reason: z.string().nullable()
})

export const CustomerProfileSubresourceResponseSchema = z.object({
  profile_status: z.enum(['READY', 'UPDATING', 'STALE', 'PARTIAL', 'FAILED', 'NOT_READY']),
  current_profile_version: z.string().nullable(),
  items: z.array(JsonRecordSchema),
  next_cursor: z.string().nullable(),
  has_more: z.boolean()
})

export const CustomerProfileResponseSchema = z.object({
  customer_id: z.string(),
  profile_status: z.enum(['READY', 'UPDATING', 'STALE', 'PARTIAL', 'FAILED', 'NOT_READY']),
  current_profile_version: z.string().nullable(),
  profile_version_number: z.number().int().nullable(),
  schema_version: z.string(),
  freshness: CustomerProfileFreshnessSchema,
  sections: CustomerProfileSectionsSchema,
  evidence_refs: z.array(CustomerProfileEvidenceRefSchema),
  links: z.object({
    changes: z.string(),
    evidence: z.string(),
    journeys: z.string(),
    follow_ups: z.string(),
    versions: z.string()
  })
})

export const CustomerProfileApiErrorSchema = z.object({
  code: z.string(),
  message: z.string(),
  details: JsonRecordSchema.nullable().optional()
})

export const CustomerProfileEnvelopeSchema = z.object({
  request_id: z.string(),
  data: z.unknown().nullable(),
  error: CustomerProfileApiErrorSchema.nullable()
})

export const CustomerProfileRefreshResponseSchema = z.object({
  request_id: z.string(),
  run_id: z.number().int(),
  profile_status: z.enum(['READY', 'UPDATING', 'STALE', 'PARTIAL', 'FAILED', 'NOT_READY']),
  current_profile_version: z.string().nullable(),
  scheduled_at: z.string()
})

export type CustomerProfileResponse = z.infer<typeof CustomerProfileResponseSchema>
export type CustomerProfileEvidence = z.infer<typeof CustomerProfileEvidenceSchema>
export type CustomerProfileRecord = z.infer<typeof JsonRecordSchema>
export type CustomerProfileRefreshResponse = z.infer<typeof CustomerProfileRefreshResponseSchema>

File: CRM-Server/app/schemas/customer_profile.py
"""Public contracts for the versioned customer profile projection."""

from __future__ import annotations

from datetime import datetime  # noqa: TC003  # Pydantic resolves this annotation at runtime.
from typing import Any, Generic, Literal, TypeVar

from pydantic import BaseModel, ConfigDict, Field

T = TypeVar("T")
CustomerProfileStatusValue = Literal["READY", "UPDATING", "STALE", "PARTIAL", "FAILED", "NOT_READY"]
CustomerProfilePublicationStatusValue = Literal[
    "DRAFT", "PUBLISHED", "PUBLISHED_WITH_WARNINGS", "SUPERSEDED", "FAILED", "REJECTED"
]


class CustomerProfileApiError(BaseModel):
    code: str
    message: str
    details: dict[str, Any] | None = None


class CustomerProfileApiEnvelope(BaseModel, Generic[T]):
    request_id: str
    data: T | None = None
    error: CustomerProfileApiError | None = None


class CustomerProfileFreshness(BaseModel):
    profile_as_of: datetime | None = None
    latest_business_event_at: datetime | None = None
    is_stale: bool = False
    stale_reason: str | None = None


class CustomerProfileSections(BaseModel):
    current_situation: dict[str, Any] = Field(default_factory=dict)
    current_journeys: list[dict[str, Any]] = Field(default_factory=list)
    important_changes: list[dict[str, Any]] = Field(default_factory=list)
    long_term_context: dict[str, Any] = Field(default_factory=dict)
    follow_up_process: list[dict[str, Any]] = Field(default_factory=list)
    recorded_follow_ups: list[dict[str, Any]] = Field(default_factory=list)


class CustomerProfileLinks(BaseModel):
    changes: str
    evidence: str
    journeys: str
    follow_ups: str
    versions: str


class CustomerProfileResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    customer_id: str
    profile_status: CustomerProfileStatusValue
    current_profile_version: str | None = None
    profile_version_number: int | None = None
    schema_version: str = "v2"
    freshness: CustomerProfileFreshness
    sections: CustomerProfileSections
    evidence_refs: list[dict[str, Any]] = Field(default_factory=list)
    links: CustomerProfileLinks


class CustomerProfileSubresourceResponse(BaseModel):
    profile_status: CustomerProfileStatusValue
    current_profile_version: str | None = None
    items: list[dict[str, Any]] = Field(default_factory=list)
    next_cursor: str | None = None
    has_more: bool = False


class CustomerProfileRefreshRequest(BaseModel):
    scope: Literal["partial", "full"] = "full"
    reason: Literal["manual_refresh", "migration", "correction"] = "manual_refresh"
    expected_current_version: int | None = Field(default=None, ge=1)


class CustomerProfileRefreshResponse(BaseModel):
    request_id: str
    run_id: int
    profile_status: CustomerProfileStatusValue
    current_profile_version: str | None = None
    scheduled_at: datetime


class CustomerProfileVersionSummary(BaseModel):
    public_id: str
    profile_version: int
    schema_version: str
    publication_status: CustomerProfilePublicationStatusValue
    source_event_key: str | None = None
    run_id: int | None = None
    graph_version: str
    source_watermark: dict[str, Any] = Field(default_factory=dict)
    fact_watermark: int = 0
    journey_watermark: int = 0
    task_watermark: int = 0
    commitment_watermark: int = 0
    generated_at: datetime
    published_at: datetime | None = None
    created_time: datetime


class CustomerProfileVersionListResponse(BaseModel):
    items: list[CustomerProfileVersionSummary]
    next_cursor: str | None = None
    has_more: bool = False
